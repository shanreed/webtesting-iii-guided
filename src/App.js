import React, { Component } from 'react';

import Speak from './components/speak';
import logo from './logo.svg';
import './App.css';







//Better Option
export const asyncFunc = () => {
  return new Promise((resolve) => { // don't forget to return!
    setTimeout(() => {
      // callback(); // looks like we will need a spy
      resolve('Success');
    }, 1000); // will this block the test for a full second??
  });
}

class App extends Component {
  state = {
    message: ''
  }
  speak = () => {
    this.setState({message: "Hello"})
  }
//avoid doing unless you have a truly static function
   static asyncFunc2 = () => {
    return new Promise((resolve) => { // don't forget to return!
      setTimeout(() => {
        // callback(); // looks like we will need a spy
        resolve('Success');
      }, 1000); // will this block the test for a full second??
    });
  }
  render() {
    return (
      <div className="App">
        <header className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
          <p>
            Edit <code>src/App.js</code> and manually reload.
          </p>
          <a
            className="App-link"
            href="https://reactjs.org"
            target="_blank"
            rel="noopener noreferrer"
          >
            Learn React
          </a>
          <Speak speak = {this.speak} message = {this.state.message}/>
        </header>
      </div>
    );
  }
}

export default App;
