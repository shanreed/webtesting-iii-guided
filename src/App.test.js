import React from 'react';
import renderer from 'react-test-renderer'; // 1: install this npm module as a dev dependency yarn add -D react-test-render
// renderer is a banana word because it is a default import
import {render, fireEvent} from '@testing-library/react';
import App, {asyncFunc} from './App';

describe('<App />', () => {
  // 2. write this test
  it('matches snapshot', () => {
    const tree = renderer.create(<App />); // generates a DOM tree

    // snapshots are a JSON representation of the DOM tree
    expect(tree.toJSON()).toMatchSnapshot();
  });
});

describe('speak', () => {
  // 2. write this test
  it('should pass "hello" into Speak ', () => {
    const {getByText, queryByText} = render(<App/>);
    // expect(queryByText(/hello/i)).toBeTruthy();//fail
    expect(queryByText(/hello/i)).toBeFalsy();//pass
    fireEvent.click(getByText(/speak/i));

    expect(queryByText(/hello/i)).toBeTruthy();

  });
});


//Testing Asynchronous code
// 3 ways:
//1st: Passing async function to describe
describe('asyncFunc', () => {
  it('eventually resolves to success', () => {
    let resolvedValue = null;
    const expected = 'Success';
   asyncFunc().then(res => {
     //assertion
    resolvedValue = res; // yup.
    expected(resolvedValue).toEqual(expected)
   })

  });
  //or
  it('eventually resolves to success 2', () => {
    // let resolvedValue = null;
    const expected = 'Success';
   asyncFunc().then(res => {
     //assertion
    // resolvedValue = res; // yup.
    expected(res).toEqual(expected)
   })
})});


//with asyncFunc inside App component, avoid doing

describe('asyncFunc2', () => {
  it('eventually resolves to success 3', () => {
    let resolvedValue = null;
    const expected = 'Success';
   App.asyncFunc2().then(res => {
     //assertion
    resolvedValue = res; // yup.
    expected(resolvedValue).toEqual(expected)
   })

  });
  //or
  it('eventually resolves to success 4', () => {
    // let resolvedValue = null;
    const expected = 'Success';
   App.asyncFunc2().then(res => {
     //assertion
    // resolvedValue = res; // yup.
    expected(res).toEqual(expected)
   })
})});

