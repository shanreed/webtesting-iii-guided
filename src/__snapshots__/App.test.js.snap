// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches snapshot 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <p>
      Edit 
      <code>
        src/App.js
      </code>
       and manually reload.
    </p>
    <a
      className="App-link"
      href="https://reactjs.org"
      rel="noopener noreferrer"
      target="_blank"
    >
      Learn React
    </a>
    <div>
      <button
        onClick={[Function]}
      >
        SPEAK
      </button>
      <div>
        
      </div>
    </div>
  </header>
</div>
`;
